### PR TITLE
Add support for JIT release mode

### PR DESCRIPTION
**Pull request by jonahwilliams**
Creates a new Flutter runtime mode "JIT_RELEASE" which implies a product vm but not a precompiled runtime. Adds a new android BuildConfig value of JIT_RELEASE

The macro changes are done. To get this working locally I had to remove a DCHECK on window size (still investigating, possibly a race condition due to slower startup).
Progress towards https://github.com/flutter/flutter/issues/9253

### CODE REVIEW

File: lib/ui/window/viewport_metrics.cc [hunk 1]
```diff
@@ -40,9 +40,9 @@ ViewportMetrics::ViewportMetrics(double p_device_pixel_ratio,
           p_physical_system_gesture_inset_bottom),
       physical_system_gesture_inset_left(p_physical_system_gesture_inset_left) {
   // Ensure we don't have nonsensical dimensions.
-  FML_DCHECK(physical_width > 0);
-  FML_DCHECK(physical_height > 0);
-  FML_DCHECK(device_pixel_ratio > 0);
+  // FML_DCHECK(physical_width > 0);
```
Comment (jonahwilliams) [inline lib/ui/window/viewport_metrics.cc]: Still working on this one

Comment (jonahwilliams) [reply lib/ui/window/viewport_metrics.cc]: Due to the slower startup, these seem to be 0 in debug JIT. Is the condition here too strong? Or does this check need to be elided for debug JIT mode?

**Comment by jonahwilliams** [comment]
cc @chingjun

File: lib/ui/window/viewport_metrics.cc [hunk 2]
```diff
@@ -40,8 +40,8 @@ ViewportMetrics::ViewportMetrics(double p_device_pixel_ratio,
           p_physical_system_gesture_inset_bottom),
       physical_system_gesture_inset_left(p_physical_system_gesture_inset_left) {
   // Ensure we don't have nonsensical dimensions.
-  FML_DCHECK(physical_width > 0);
-  FML_DCHECK(physical_height > 0);
+  FML_DCHECK(physical_width >= 0);
```
Comment (jason-simmons) [inline lib/ui/window/viewport_metrics.cc]: Is this related to jit_release mode?

File: runtime/dart_snapshot.cc [hunk 3]
```diff
@@ -23,8 +23,10 @@ const char* DartSnapshot::kIsolateInstructionsSymbol =
 // On Windows and Android (in debug mode) the engine finds the Dart snapshot
 // data through symbols that are statically linked into the executable.
 // On other platforms this data is obtained by a dynamic symbol lookup.
-#define DART_SNAPSHOT_STATIC_LINK \
-  (OS_WIN || (OS_ANDROID && FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG))
+#define DART_SNAPSHOT_STATIC_LINK                                        \
+  (OS_WIN ||                                                             \
+   (OS_ANDROID && (FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG || \
```
Comment (jason-simmons) [inline runtime/dart_snapshot.cc]: I think it would be clearer to define a macro that is set whenever the engine is built with a JIT runtime.

It might also be useful to create a macro for modes that are "release modes".  I suspect there are other places where the engine currently checks for `FLUTTER_RUNTIME_MODE_RELEASE` that should be updated to include `JIT_RELEASE` (e.g. `shell/common/switches.cc`)

These can be defined by a new header in `flutter/common` based on the `FLUTTER_RUNTIME_MODE` set by the build scripts.

File: lib/ui/window/viewport_metrics.cc (hunk 2 above)
Comment (jonahwilliams) [reply lib/ui/window/viewport_metrics.cc]: Yes, see https://github.com/flutter/engine/pull/12446#discussion_r328337326

File: lib/ui/window/viewport_metrics.cc (hunk 1 above)
Comment (jonahwilliams) [reply lib/ui/window/viewport_metrics.cc]: EDIT: I mean release JIT

File: runtime/dart_snapshot.cc (hunk 3 above)
Comment (jonahwilliams) [reply runtime/dart_snapshot.cc]: Do you have any preference towards a positive/negative approach on these macros? For example PRECOMPILED_RUNTIME (which matches the dart_vm getter) vs JIT_RUNTIME (which calls out the unique behavior) ?

File: tools/gn [hunk 4]
```diff
@@ -152,6 +152,8 @@ def to_gn_args(args):
         gn_args['dart_runtime_mode'] = 'profile'
     elif runtime_mode == 'dynamic_release':
         gn_args['dart_runtime_mode'] = 'release'
+    elif runtime_mode == 'jit_release':
+        gn_args['dart_runtime_mode'] = 'release';
```
Comment (chinmaygarde) [inline tools/gn]: `jit_release` so both variants can in the out directory and engineers tinker on builds locally without one clobbering another.

File: lib/ui/window/viewport_metrics.cc (hunk 2 above)
Comment (chinmaygarde) [reply lib/ui/window/viewport_metrics.cc]: I think these checks are just too strong and it should be possible (and often is on release builds) that the viewport metrics dimensions can be zero. Let's just get rid of them.

File: runtime/dart_snapshot.cc (hunk 3 above)
Comment (jason-simmons) [reply runtime/dart_snapshot.cc]: I'd vote for `FLUTTER_JIT_RUNTIME` and `FLUTTER_RELEASE` as names for these macros

File: lib/ui/window/viewport_metrics.cc (hunk 2 above)
Comment (jonahwilliams) [reply lib/ui/window/viewport_metrics.cc]: I've removed all of them

File: tools/gn (hunk 4 above)
Comment (jonahwilliams) [reply tools/gn]: dart_runtime_mode is used by the Dart SDK. Otherwise, the output directory I have locally is host_jit_release(_unopt) and android_jit_release(_unopt) so this should be okay?

File: shell/platform/android/flutter_main.cc [hunk 5]
```diff
@@ -102,7 +103,7 @@ void FlutterMain::Init(JNIEnv* env,
     fml::MessageLoop::GetCurrent().RemoveTaskObserver(key);
   };
 
-#if FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG
+#if FLUTTER_JIT_RUNTIME
```
Comment (zanderso) [inline shell/platform/android/flutter_main.cc]: ditto

File: shell/platform/android/flutter_main.cc [hunk 6]
```diff
@@ -26,7 +27,7 @@
 namespace flutter {
 
 extern "C" {
-#if FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG
+#if FLUTTER_JIT_RUNTIME
 // Used for debugging dart:* sources.
```
Comment (zanderso) [inline shell/platform/android/flutter_main.cc]: Is this comment accurate? If it is, then debugging dart:* sources in JIT_RELEASE doesn't seem right.

File: shell/platform/embedder/embedder.cc [hunk 7]
```diff
@@ -15,11 +15,11 @@
 #endif  // OS_WIN
 
 extern "C" {
-#if FLUTTER_RUNTIME_MODE == FLUTTER_RUNTIME_MODE_DEBUG
+#if FLUTTER_JIT_RUNTIME
 // Used for debugging dart:* sources.
```
Comment (zanderso) [inline shell/platform/embedder/embedder.cc]: Same comment here as above.

File: tools/gn [hunk 8]
```diff
@@ -152,6 +152,8 @@ def to_gn_args(args):
         gn_args['dart_runtime_mode'] = 'profile'
     elif runtime_mode == 'dynamic_release':
```
Comment (zanderso) [inline tools/gn]: Are 'dynamic_profile' and 'dynamic_release' still used anywhere?

Comment (jonahwilliams) [reply tools/gn]: No, I think these could be cleaned up here too if that's what you're asking :)

File: shell/platform/android/flutter_main.cc (hunk 6 above)
Comment (jonahwilliams) [reply shell/platform/android/flutter_main.cc]: Ahh you're right. I think I confused this with the application kernel file, but it is just the SDK sources and can be left out

File: tools/gn (hunk 8 above)
Comment (jonahwilliams) [reply tools/gn]: Removed

File: shell/platform/embedder/embedder.cc (hunk 7 above)
Comment (jonahwilliams) [reply shell/platform/embedder/embedder.cc]: Fixed

File: shell/platform/android/flutter_main.cc (hunk 5 above)
Comment (jonahwilliams) [reply shell/platform/android/flutter_main.cc]: Fixed

File: runtime/dart_snapshot.cc (hunk 3 above)
Comment (jonahwilliams) [reply runtime/dart_snapshot.cc]: I've updated them where appropriate to use `FLUTTER_RELEASE` and `FLUTTER_JIT_RUNTIME`

File: tools/gn (hunk 4 above)
Comment (chinmaygarde) [reply tools/gn]: My bad. I misread this line to mean it was the out directory. Thanks for the clarification.

**Comment by jonahwilliams** [comment]
Update: tests are all passing locally - turns out I messed up the macro definitions! _how embarassing_
